spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,223 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.source;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.sources.v2.DataSourceOptions;
+import org.apache.spark.sql.sources.v2.reader.streaming.MicroBatchReader;
+import org.apache.spark.sql.sources.v2.reader.streaming.Offset;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StreamingReader extends Reader implements MicroBatchReader {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingReader.class);
+
+  private Offset startOffset;
+  private Offset endOffset;
+
+  private final Table table;
+  private final int maxTasksPerTrigger;
+
+  static class SnapshotTasks {
+    private final long snapshotId;
+    private final List<CombinedScanTask> tasks;
+    private final int size;
+
+    @VisibleForTesting
+    SnapshotTasks(long snapshotId, List<CombinedScanTask> tasks) {
+      this.snapshotId = snapshotId;
+      this.tasks = tasks;
+      this.size = tasks.size();
+    }
+
+    long snapshotId() {
+      return snapshotId;
+    }
+
+    List<CombinedScanTask> tasks() {
+      return tasks;
+    }
+
+    int size() {
+      return size;
+    }
+  }
+
+  StreamingReader(Table table, Broadcast<FileIO> io, Broadcast<EncryptionManager> encryptionManager,
+                  boolean caseSensitive, DataSourceOptions options) {
+    super(table, io, encryptionManager, caseSensitive, options);
+
+    this.table = table;
+    this.maxTasksPerTrigger =
+      options.get("max-tasks-per-trigger").map(Integer::parseInt).orElse(1000);
+    Preconditions.checkArgument(this.maxTasksPerTrigger > 0,
+        "Option max-tasks-per-trigger should > 0");
+  }
+
+  @Override
+  public void setOffsetRange(Optional<Offset> start, Optional<Offset> end) {
+    startOffset = start.orElse(calculateInitialOffset());
+    endOffset = Optional.ofNullable(startOffset)
+        .map(s -> end.orElse(calculateEndOffset((StreamingOffset) s)))
+        .orElse(null);
+  }
+
+  @Override
+  public Offset getStartOffset() {
+    if (startOffset == null) {
+      throw new IllegalStateException("Start offset is not set");
+    }
+
+    return startOffset;
+  }
+
+  @Override
+  public Offset getEndOffset() {
+    if (endOffset == null) {
+      throw new IllegalStateException("End offset is not set");
```

Comment:
Why set `endOffset` to null when setting offsets if it is not allowed to be null?